components/places/src/storage.rs

```diff
@@ -303,6 +303,60 @@ fn add_visit(
     Ok(RowId(rid))
 }
 
+/// Returns the GUID for the specified Url, or None if it doesn't exist.
+pub fn url_to_guid(db: &impl ConnExt, url: &Url) -> Result<Option<SyncGuid>> {
+    let sql = "SELECT guid FROM moz_places WHERE url_hash = hash(:url) AND url = :url";
+    let result: Option<(SyncGuid)> = db.try_query_row(
+        sql,
+        &[(":url", &url.clone().into_string())],
+        // subtle: we explicitly need to specify rusqlite::Result or the compiler
+        // struggles to work out what error type to return from try_query_row.
+        |row| -> rusqlite::Result<_> { Ok(row.get_checked::<_, SyncGuid>(0)?) },
+        true,
+    )?;
+    Ok(result)
+}
+
+/// Internal function for deleting a place, creating a tombstone if necessary.
+/// Assumes a transaction is already set up by the caller.
+fn do_delete_place_by_guid(db: &impl ConnExt, guid: &SyncGuid) -> Result<()> {
+    // sadly we need to do a read first - we only create tombstones for
+    // history with sync_status == SyncStatus::Normal
+    let sql = "SELECT id, sync_status FROM moz_places WHERE guid = :guid";
+    let result: Option<(RowId, SyncStatus)> = db.try_query_row(
+        sql,
+        &[(":guid", guid)],
+        // subtle: we explicitly need to specify rusqlite::Result or the compiler
+        // struggles to work out what error type to return from try_query_row.
+        |row| -> rusqlite::Result<_> {
+            Ok((
+                row.get_checked::<_, RowId>(0)?,
+                SyncStatus::from_u8(row.get_checked::<_, u8>(1)?),
+            ))
+        },
+        true,
+    )?;
+    let (id, status) = match result {
+        None => return Ok(()),
+        Some(tuple) => tuple,
+    };
+    if status == SyncStatus::Normal {
```

Comment:
I think you could write this as:

```sql
INSERT OR IGNORE INTO moz_places_tombstones (guid)
SELECT guid FROM moz_places
WHERE guid = :guid AND sync_status = :status
```

That'll push the read down into SQLite, and also do nothing if the GUID doesn't exist, or isn't synced.